### PR TITLE
Explore: Fix for url does not keep query after browser refresh

### LINKED ISSUE #15862: Explore url does not keep query
**What happened**:
When copying the URL generated by Explore and loading the page, the query field is empty.

**What you expected to happen**:
The query field should be filled with the query retrieved from the url

**How to reproduce it (as minimally and precisely as possible)**:
1. Go to Explore
2. Select Graphite datasource
3. Build query
4. Hit refresh (browser)

**Environment**:
- Grafana version: 6.0.0
- Data source type & version: Graphite 0.9.16
- OS Grafana is installed on: CentOS 7.5
- User OS & Browser: MacOS Safari

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This PR makes sure that a pasted URL in Explore also considers `target` or `datasource` as metric property, before this PR only `expr` was consider a metric property.

**Which issue(s) this PR fixes**:
Fixes: #15862

### CODE REVIEW

File: public/app/core/utils/explore.ts [hunk 1]
```diff
@@ -157,49 +157,72 @@ export function buildQueryTransaction(
 
 export const clearQueryKeys: (query: DataQuery) => object = ({ key, refId, ...rest }) => rest;
 
-const isMetricSegment = (segment: { [key: string]: string }) => segment.hasOwnProperty('expr');
+const metricProperties = ['expr', 'target'];
```
Comment (torkelo) [inline public/app/core/utils/explore.ts]: this seems very specific to Prometheus and Graphite, a query model can contain any nested json object.

**Comment by torkelo** [comment]
This approach won't work for other data sources, like elasticsearch, testdata, etc that do not use target property in the query model.

**Comment by hugohaggmark** [comment]
I understand, I thought the work to use various data sources in Explore as elastic was to be done in 6.2. So this is a fix for 6.1 and the rest is a larger refactor I believe. Maybe we could sync about this, I'm not familiar with all the different query models and how to resolve them from the url.

**Comment by torkelo** [comment]
Every datasource works in explore in 6.0 just the logs part that is only Loki specific at this point

**Comment by hugohaggmark** [comment]
So adding `datasource`to the list should solve most datasources. I've Tried with Elastic, Influx, Postgres, MySql and TestData. Is that an `ok` solution or do we have a more intricate way of determine this?

**Comment by torkelo** [comment]
yes, that looks like it will work!

**Comment by hugohaggmark** [comment]
I'll fix that and squash and force push changes to branch.

**Comment by torkelo** [comment]
not sure we need to cherry pick this, release in 6.1 should be enough.

File: public/app/core/utils/explore.ts [hunk 2]
```diff
@@ -157,49 +157,72 @@ export function buildQueryTransaction(
...
+  RangeTo = 1,
+  Datasource = 2,
+  SegmentsStart = 3,
+}
+
+enum ParseUiStateIndex {
+  Graph = 0,
+  Logs = 1,
+  Table = 2,
+  Strategy = 3,
```
Comment (davkal) [inline public/app/core/utils/explore.ts]: Probably prefix this with `LogDedup`

File: public/app/core/utils/explore.ts [hunk 3]
```diff
@@ -157,49 +157,72 @@ export function buildQueryTransaction(
...
+const metricProperties = ['expr', 'target', 'datasource'];
+const isMetricSegment = (segment: { [key: string]: string }) =>
+  metricProperties.some(prop => segment.hasOwnProperty(prop));
 const isUISegment = (segment: { [key: string]: string }) => segment.hasOwnProperty('ui');
 
+enum ParseUrlStateIndex {
+  RangeFrom = 0,
+  RangeTo = 1,
+  Datasource = 2,
+  SegmentsStart = 3,
```
Comment (davkal) [inline public/app/core/utils/explore.ts]: Bit unclear what the semantics of "segment" is.

**Review by davkal — COMMENTED** [review]
2 small nits. Otherwise LGTM

File: public/app/core/utils/explore.ts (hunk 3 above)
Comment (hugohaggmark) [reply public/app/core/utils/explore.ts]: What about QuerySegementsStart?

Comment (hugohaggmark) [reply public/app/core/utils/explore.ts]: Changed to just `Queries`

**Pull request merged**